### ISSUE TITLE
Network only supports action_specs with shape in [(), (1,)])   In call to configurable 'QNetwork' 

### ISSUE THREAD

**Issue by komal-SkyNET** [tensorflow/agents#329]
Google colab here: https://github.com/komal-SkyNET/ai-neural-networks/blob/master/google-colab/dqn_tf_nes_gym.ipynb

**Comment by komal-SkyNET** [comment]
```
[BoundedTensorSpec(shape=(12,), dtype=tf.int8, name='action', minimum=array(0, dtype=int8), maximum=array(1, dtype=int8))]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-48-1f97b7ebd60b> in <module>()
      6     train_env.observation_spec(),
      7     train_env.action_spec(),
----> 8     fc_layer_params=fc_layer_params)

6 frames
/usr/local/lib/python3.6/dist-packages/tf_agents/networks/q_network.py in validate_specs(***failed resolving arguments***)
     37   if flat_action_spec[0].shape not in [(), (1,)]:
     38     raise ValueError(
---> 39         'Network only supports action_specs with shape in [(), (1,)])')
     40 
     41 

ValueError: Network only supports action_specs with shape in [(), (1,)])
  In call to configurable 'QNetwork' (<function QNetwork.__init__ at 0x7f7ef48ad950>)
```

**Comment by kbanoop** [comment]
DQN doesnt currently support more than one action: https://github.com/tensorflow/agents/blob/3ef63ca150b61d5aecce0abf289fbf0992c4f046/tf_agents/agents/dqn/dqn_agent.py#L265

Contributions are welcome. 

A workaround is to wrap the environment and change the actions to be one dimensional, e.g. if you have 2 actions each in {0, 1}, you could turn that into one action taking values in {0, 3} where {0} maps to {0, 0}, {1} maps to {0, 1}, {2} maps to {1, 0} and {3} maps to {1, 1}.

**Comment by vinayjoshi22** [comment]
> DQN doesnt currently support more than one action:
> 
> https://github.com/tensorflow/agents/blob/3ef63ca150b61d5aecce0abf289fbf0992c4f046/tf_agents/agents/dqn/dqn_agent.py#L265
> 
> Contributions are welcome.
> 
> A workaround is to wrap the environment and change the actions to be one dimensional, e.g. if you have 2 actions each in {0, 1}, you could turn that into one action taking values in {0, 3} where {0} maps to {0, 0}, {1} maps to {0, 1}, {2} maps to {1, 0} and {3} maps to {1, 1}.

This is a good idea however since the output dimension of the Q-network is (1,) I suspect Q-network would not learn well since regression over an output variable of large value could cause problems in generalization.

Also, this limit on the size of action_spec is it because of DQN implementation in TF only supports actions_spec size 1 or only Q-network supports action_spec size (1,). I do not understand where this limit is coming from. Any explanation in the matter is appreciated.

**Comment by vinayjoshi22** [comment]
> > DQN doesnt currently support more than one action:
> > https://github.com/tensorflow/agents/blob/3ef63ca150b61d5aecce0abf289fbf0992c4f046/tf_agents/agents/dqn/dqn_agent.py#L265
> > 
> > Contributions are welcome.
> > A workaround is to wrap the environment and change the actions to be one dimensional, e.g. if you have 2 actions each in {0, 1}, you could turn that into one action taking values in {0, 3} where {0} maps to {0, 0}, {1} maps to {0, 1}, {2} maps to {1, 0} and {3} maps to {1, 1}.
> 
> This is a good idea however since the output dimension of the Q-network is (1,) I suspect Q-network would not learn well since regression over an output variable of large value could cause problems in generalization.
> 
> Also, this limit on the size of action_spec is it because of DQN implementation in TF only supports actions_spec size 1 or only Q-network supports action_spec size (1,). I do not understand where this limit is coming from. Any explanation in the matter is appreciated.

I just saw [here](https://github.com/tensorflow/agents/blob/eb3bd090d6c6b4e601cee9be19ff0e834e6fe7bf/tf_agents/agents/dqn/dqn_agent.py#L266) that DQN algorithm implementation in TF only supports one-dimensional action_spec.  I would like to see TF support for multi-dimensional action_spec.

**Comment by JThFr** [comment]
Is there currently a wrapper to transform multi-dimensional action_spec into one-dimensional action_spec?
If not I might try to do it.

**Comment by darkmatter2222** [comment]
> Is there currently a wrapper to transform multi-dimensional action_spec into one-dimensional action_spec?
> If not I might try to do it.

https://github.com/tensorflow/agents/issues/97
https://www.tensorflow.org/agents/tutorials/2_environments_tutorial
https://www.tensorflow.org/agents/api_docs/python/tf_agents/environments/wrappers/ActionDiscretizeWrapper

**Comment by clockzhong** [comment]
@JThFr , I don't think it's a good idea to translate a multi-dimensional action_spec to a one-dimensional action_spec. As @vinayjoshi22 explained above, the observation data doesn't relate to this single value, it's much much more difficult to find the connections between the observation data and the human-made one-dimensional action choices than the multi-dimensional action choices directly. 
Of course, if you've implemented it, and tried it, I'm glad to learn your test result, it's also possible that the neural network feels very easy to find those data connections than we human. But I still feel we need enable multi-dimensional action-spec, it's elegance on the system design.

**Comment by ebrevdo** [comment]
You can now use your own sequence of Keras layers - one that can emit whatever action distribution you want.  See `tf_agents.network.Sequential` and unit tests.

**Comment by ebrevdo** [comment]
(but DQN is based on a paper that did not consider larger action spaces; and we're trying to faithfully reproduce that algorithm)

**Comment by alexhiggins732** [comment]
> You can now use your own sequence of Keras layers - one that can emit whatever action distribution you want. See `tf_agents.network.Sequential` and unit tests.

Using tf-agents nightly, and it still doesn't work. Unit tests still assert [only scalar actions](https://github.com/tensorflow/agents/blob/58ca29e1c57332bd34d72f9b1387e23743e461c8/tf_agents/agents/dqn/dqn_agent_test.py#L156).

**Comment by ebrevdo** [comment]
What action type do you want and what dqn algorithm that's published can
handle your action type?

**Comment by alexhiggins732** [comment]
So two different questions: Being able to use TF-Agents in the Gym multi-discrete environments would be nice.

Personally, I spent some time building a PC mouse driver environment for tf-agent training. Theoretically, an agent should be able to use the environment as an interface to against arbitrary environments using a single set of input controls. Originally, I had a Discrete(8) for the mouse buttons and 2 continuous floats for the mouse speed and and angle. TF-Agents didn't like that, s

I then converted it over to 3 Discrete(8) actions, assuming the DQN TF-agent might play nice with the Gym Multi-Discrete:

> Mouse Buttons:  Touch, LeftButton, MiddleButton,RightButton, ScrollUp,ScrollDown, IncreaseScrollSpeed,DecreseScroll Speed
> Direction: Each represents an Angle 0,45.. 315
> Speed: Each represents a speed used as an exponent 2^N

However, it does not.

I could just convert this into a Discrete(24) but I can envision numerous scenarios that benefit from the allowing multiple actions. Not to mention all of the additional state management in the environment. Not on for this scenario, but for other future projects I have planned.

As a side note, DQN using a `q_network.QNetwork` "runs" against the [Gym Taxi Environment](https://github.com/tw2665/reinforcement_learning_taxi/blob/master/Deep_Q_Network_Taxi.ipynb) - although it doesn't seem good at solving it or even exploring.

However, using the latest [DQN TF-Agent tutorial](https://www.tensorflow.org/agents/tutorials/1_dqn_tutorial)  - which is ran against cartpole using the sequential - blows up when trying to run it against the Gym taxi environment, which also takes scalar input:

`ValueError: Input 0 of layer dense is incompatible with the layer: : expected min_ndim=2, found ndim=1. Full shape received: (1,)`
